app/services/admin/charts_data.rb

```diff
@@ -0,0 +1,33 @@
+module Admin
+  class ChartsData
+    def initialize(length = 7)
+      @length = length
+    end
+
+    def call
+      period = @length.days.ago..Time.current
+      previous_period = (@length * 2).days.ago..@length.days.ago
+
+      grouped_posts = Article.where(published_at: period).group("DATE(published_at)").size
+      grouped_comments = Comment.where(created_at: period).group("DATE(created_at)").size
+      grouped_reactions = Reaction.where(created_at: period).group("DATE(created_at)").size
+      grouped_users = User.where(registered_at: period).group("DATE(registered_at)").size
+
+      posts_values = (0..@length - 1).map { |n| grouped_posts[n.days.ago.to_date] || 0 }.reverse
+      comments_values = (0..@length - 1).map { |n| grouped_comments[n.days.ago.to_date] || 0 }.reverse
+      reactions_values = (0..@length - 1).map { |n| grouped_reactions[n.days.ago.to_date] || 0 }.reverse
+      new_members_values = (0..@length - 1).map { |n| grouped_users[n.days.ago.to_date] || 0 }.reverse
```

Comment:
Suggestions (non-blocking):

1. Don't create the same range 4 times in a row. 
2. Don't `map` + `reverse`, use a different iteration method instead.

```suggestion
      days_range = (@length - 1).downto(0)
      posts_values = days_range.map { |n| grouped_posts[n.days.ago.to_date] || 0 }
      comments_values = days_range.map { |n| grouped_comments[n.days.ago.to_date] || 0 }
      reactions_values = days_range.map { |n| grouped_reactions[n.days.ago.to_date] || 0 }
      new_members_values = days_range.map { |n| grouped_users[n.days.ago.to_date] || 0 }
```

Reply:
Thanks!